man/schroot.1.in
.\" Copyright © 2005-2010  Roger Leigh <rleigh@debian.org>
.\" Copyright © 2006       Andreas Bombe <aeb@debian.org>
.\"
.\" schroot is free software: you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" schroot is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see
.\" <http://www.gnu.org/licenses/>.
.\"
.TH SCHROOT 1 "@RELEASE_DATE_S@" "Version @VERSION@" "Debian sbuild"
.SH NAME
schroot \- securely enter a chroot environment
.SH SYNOPSIS
.B schroot
.RB [ \-h \[or] \-\-help " \[or] " \-V \[or] \-\-version
.RB " \[or] " \-l \[or] \-\-list " \[or] " \-i \[or] \-\-info
.RB " \[or] " \-\-config " \[or] " \-\-location
.RB " \[or] " \-\-automatic\-session
.RB " \[or] " \-b \[or] \-\-begin\-session " \[or] " \-\-recover\-session
.RB " \[or] " \-r \[or] \-\-run\-session " \[or] " \-e \[or] \-\-end\-session ]
.RB [ \-f \[or] "\-\-force" ]
.RB [ "\-n \fIsession-name\fP" \[or] "\-\-session\-name=\fIsession-name\fP" ]
.RB [ "\-d \fIdirectory\fP" \[or] "\-\-directory=\fIdirectory\fP" ]
.RB [ "\-u \fIuser\fP" \[or] "\-\-user=\fIuser\fP" ]
.RB [ \-p \[or] \-\-preserve\-environment ]
.RB [ \-q \[or] \-\-quiet " \[or] " \-v \[or] \-\-verbose ]
.RB [ "\-c \fIchroot\fP" \[or] "\-\-chroot=\fIchroot\fP"
.RB " \[or] " \-\-all " \[or] " \-\-all\-chroots " \[or] " \-\-all\-source\-chroots " \[or] " \-\-all\-sessions ]
.RB [ \-\- ]
.RB [ COMMAND " [ " ARG1 " [ " ARG2 " [ " ARGn ]]]]
.SH DESCRIPTION
\fBschroot\fP allows the user to run a command or a login shell in a chroot
environment.  If no command is specified, a login shell will be started in the
user's current working directory inside the chroot.
.PP
The command is a program, plus as many optional arguments as required.  Each
argument may be separately quoted.
.PP
The directory the command or login shell is run in depends upon the context.
See \fI\-\-directory\fP option below for a complete description.
.PP
All chroot usage will be logged in the system logs.  Under some circumstances,
the user may be required to authenticate themselves; see the section
\[lq]\fIAuthentication\fP\[rq], below.
.PP
If no chroot is specified, the chroot name or alias \[oq]default\[cq] will be
used as a fallback.  This is equivalent to \[lq]\-\-chroot=default\[rq].
.SH OVERVIEW
There is often a need to run programs in a virtualised environment rather than
on the host system directly.  Unlike other virtualisation systems such as
\fBkvm\fP or \fPXen\fP, schroot does not virtualise the entire system; it only
virtualises the filesystem, and some parts of the filesystem may still be
shared with the host.  It is therefore fast, lightweight and flexible.
However, it does not virtualise other aspects of the system, such as shared
memory, networking, devices etc., and so may be less secure than other systems,
depending upon its intended use.  Some examples of existing uses for schroot
include:
.IP \[bu]
Running an untrusted program in a sandbox, so that it can't interfere with
files on the host system; this may also be used to limit the damage a
compromised service can inflict upon the host
.IP \[bu]
Using a \fIdefined\fP or \fIclean\fP environment, to guarantee the
reproducibility and integrity of a given task
.IP \[bu]
Using different versions of an operating system, or even different operating
systems altogether, e.g. different GNU/Linux distributions
.IP \[bu]
Running 32-bit programs using a 32-bit chroot on a 64-bit host system
.IP \[bu]
Automatic building of Debian packages using
.BR sbuild (1),
which builds each package in a pristine chroot snapshot when using LVM
snapshots or unions
.IP \[bu]
Supporting multiple system images in a cluster setup, where modifying the base
image is time-consuming and/or supporting all the required configurations
needed by users is difficult: different chroots can support all the different
configurations required, and cluster users may be given access to the chroots
they need (which can include root access for trusted users to maintain their
own images)
.PP
A chroot may be used directly as root by running
.BR chroot (8),
but normal users are not able to use this command.  \fBschroot\fP allows access
to chroots for normal users using the same mechanism, but with permissions
checking and allowing additional automated setup of the chroot environment,
such as mounting additional filesystems and other configuration tasks.  This
automated setup is done through the action of \fIsetup scripts\fP which may be
customised and extended to perform any actions required.
.SH OPTIONS
\fBschroot\fP accepts the following options:
.SS Actions
.TP
.BR \-h ", " \-\-help
Show help summary.
.TP
.BR \-V ", " \-\-version
Print version information.
.TP
.BR \-l ", " \-\-list
List all available chroots.
.TP
.BR \-i ", " \-\-info
Print detailed information about the specified chroots.
.TP
.BR \-\-config
Print configuration of the specified chroots.  This is useful for testing that
the configuration in use is the same as the configuration file.  Any comments
in the original file will be missing.
.TP
.BR \-\-location
Print location (path) of the specified chroots.  Note that chroot types which
can only be used within a session will not have a location until they are
active.
.SS General options
.TP
.BR \-q ", " \-\-quiet
Print only essential messages.
.TP
.BR \-v ", " \-\-verbose
Print all messages.
.SS Chroot selection
.TP
.BR \-c ", " \-\-chroot=\fIchroot\fP
Specify a chroot or active session to use.  This option may be used multiple
times to specify more than one chroot, in which case its effect is similar to
\fI\-\-all\fP.  The chroot name may be prefixed with a \fInamespace\fP; see the
section \[lq]\fIChroot Namespaces\fP\[rq], below.
.TP
.BR \-a ", " \-\-all
Select all chroots, source chroots and active sessions.  When a command has
been specified, the command will be run in all chroots, source chroots and
active sessions.  If \fI\-\-info\fP has been used, display information about
all chroots.  This option does not make sense to use with a login shell (when
no command has been specified).  This option is equivalent to
\[lq]\-\-all\-chroots \-\-all\-source\-chroots \-\-all\-sessions\[rq].
.TP
.BR \-\-all\-chroots
Select all chroots.  Identical to \fI\-\-all\fP, except that source chroots and
active sessions are not considered.
.TP
.BR \-\-all\-sessions
Select all active sessions.  Identical to \fI\-\-all\fP, except that chroots
and source chroots are not considered.
.TP
.BR \-\-all\-source\-chroots
Select all source chroots.  Identical to \fI\-\-all\fP, except that chroots
and sessions are not considered.
.SS Chroot environment
.TP
.BR \-d ", " \-\-directory=\fIdirectory\fP
Change to \fIdirectory\fP inside the chroot before running the command or login
shell.  If \fIdirectory\fP is not available, schroot will exit with an error
status.
.IP
The default behaviour is as follows (all directory paths are inside the
chroot).  A login shell is run in the current working directory.  If this is
not available, it will try $HOME (when \fI\-\-preserve\-environment\fP is
used), then the user's home directory, and \fI/\fP inside the chroot in turn.
A command is always run in the current working directory inside the chroot.  If
none of the directories are available, schroot will exit with an error status.
.TP
.BR \-u ", " \-\-user=\fIuser\fP
Run as a different user.  The default is to run as the current user.  If
required, the user may be required to authenticate themselves with a password.
For further information, see the section \[lq]\fIAuthentication\fP\[rq], below.
.TP
.BR \-p ", " \-\-preserve\-environment
Preserve the user's environment inside the chroot environment.  The default is
to use a clean environment; this option copies the entire user environment and
sets it in the session.  The environment variables allowed are subject to
certain restrictions; see the section \[lq]\fIEnvironment\fP\[rq], below.
.SS Session actions
.TP
.BR \-\-automatic\-session
Begin, run and end a session automatically.  This is the default action, so
does not require specifying in normal operation.
.TP
.BR \-b ", " \-\-begin\-session
Begin a session.  A unique session identifier (session ID) is returned on
standard output.  The session ID is required to use the other session options.
Note that the session identifier may be specified with the
\fI\-\-session\-name\fP option.
.TP
.BR \-\-recover\-session
Recover an existing session.  If an existing session has become unavailable,
for example becoming unmounted due to a reboot, this option will make the
session available for use again, for example by remounting it.  The session ID
is specified with the \fI\-\-chroot\fP option.
.TP
.BR \-r ", " \-\-run\-session
Run an existing session.  The session ID is specified with the \fI\-\-chroot\fP
option.
.TP
.BR \-e ", " \-\-end\-session
End an existing session.  The session ID is specified with the \fI\-\-chroot\fP
option.
.SS Session options
.TP
.BR \-n ", " \-\-session\-name=\fIsession-name\fP
Name a session.  The specified \fIsession-name\fP replaces the default session
name containing an automatically-generated session ID.  The session name must
not contain a namespace qualifier, since sessions are always created within the
\[oq]session:\[cq] namespace.  The session name is also subject to the chroot
naming restrictions documented in
.BR schroot.conf (5).
.TP
.BR \-f ", " \-\-force
Force a session operation, even if it would otherwise fail.  This may be used
to forcibly end a session, even if it has active users.  This does not
guarantee that the session will be ended cleanly; filesystems may not be
unmounted, for example.
.SS Separator
.TP
.BR \-\-
End of options.  Used to indicate the end of the schroot options; any following
options will be passed to the command being run, rather than to schroot.
.SH AUTHENTICATION
If the user is not an allowed user, or a member of the allowed groups (or if
changing to root, the allowed root users or allowed root groups) for the
specified chroot(s), permission will be immediately denied.  If switching
users, and the user running the command has access, the user will be required
to authenticate themselves using the credentials of the user being switched to.
.PP
On systems supporting Pluggable Authentication Modules (PAM), schroot will use
PAM for authentication and authorisation of users.  If and when required,
schroot will prompt for a password.  If PAM is not available, all
authentication will automatically fail (user switching is \fInot\fP supported
without PAM).
.PP
Note that when PAM is in use, the root user is not granted any special
privileges by default in the program.  However, the default PAM configuration
permits root to log in without a password (\fIpam_rootok.so\fP), but this may
be disabled to prevent root from accessing any chroots except if specifically
permitted.  In such a situation, root must be added to the allowed users or
groups as for any other user or group.  If PAM is not available, the root user
will be permitted to access all chroots, even when not explicitly granted
access.
.SH CHROOT NAMESPACES
.SS Namespace basics
There are three different types of chroot: regular chroots, source chroots and
session chroots.  These different types of chroot are separated into different
\fInamespaces\fP.  A namespace is a prefix to a chroot name.  Currently there
are three namespaces: \[oq]chroot:\[cq], \[oq]source:\[cq] and
\[oq]session:\[cq].  Use \fI\-\-list \-\-all\fP to list all available chroots
in all namespaces.  Because \[oq]:\[cq] is used as the separator between
namespace and chroot names, it is not permitted to use this character in chroot
names.
.PP
Depending upon the action you request schroot to take, it may look for the
chroot in one of the three namespaces, or a particular namespace may be
specified.  For example, a chroot named \[lq]sid\[rq] is actually named
\[lq]chroot:sid\[rq] if the namespace is included, but the namespace may be
omitted for most actions.
.SS Source chroots
Some chroot types, for example LVM snapshots and Btrfs snapshots, provide
session-managed copy-on-write snapshots of the chroot.  These also provide a
\fIsource chroot\fP to allow easy access to the filesystem used as a source for
snapshotting.  These are regular chroots as well, just with the snapshotting
disabled.  For a chroot named \[lq]sid\-snapshot\[rq] (i.e. with a fully
qualified name of \[lq]chroot:sid\-snapshot\[rq]), there will also be a
corresponding source chroot named \[lq]source:sid\-snapshot\[rq].  Earlier
versions of schroot provided source chroots with a \[oq]\-source\[cq] suffix.
These are also provided for compatibility.  In this example, this would be
called \[lq]chroot:sid\-snapshot\-source\[rq].  These compatibility names will
be dropped in a future version, so programs and scripts should switch to using
the namespace-qualified names rather than the old suffix.
.SS Session chroots
All sessions created with \fI\-\-begin\-session\fP are placed within the
\[oq]session:\[cq] namespace.  A session named with \fI\-\-session\-name\fP may
have any name, even the same name as the chroot it was created from, providing
that it is unique within this namespace.  This was not permitted in previous
versions of schroot which did not have namespaces.
.SS Actions and default namespaces
All actions use \[oq]chroot:\[cq] as the default namespace, with some session
actions being the exception.  \fI\-\-run\-session\fP,
\fI\-\-recover\-session\fP and \fI\-\-end\-session\fP use \[oq]session:\[cq] as
the default namespace instead, since these actions work on session chroots.
The upshot is that the namespace is usually never required except when you need
to work with a chroot in a namespace other than the default, such as when using
a source chroot.  To make chroot selection unambiguous, it is always possible
to use the full name including the namespace, even when not strictly required.
.SH PERFORMANCE
.PP
Performance on some filesystems, for example Btrfs, is bad when running dpkg
due to the amount of fsync operations performed.  This may be mitigated by
installing the eatmydata package and then adding eatmydata to the
\f[CI]command\-prefix\fP configuration key, which disables all fsync
operations.  Note that this should only be done in snapshot chroots where data
loss is not an issue.  This is useful when using a chroot for package building,
for example.
.SH EXAMPLES
.SS List available chroots
\f[CR]%\ \f[CB]schroot \-l\fP\fP\[CR]
.br
\f[CR]chroot:default\fP
.br
\f[CR]chroot:etch\fP
.br
\f[CR]chroot:sid\fP
.br
\f[CR]chroot:testing\fP
.br
\f[CR]chroot:unstable\fP
.SS Get information about a chroot
\f[CR]%\ \f[CB]schroot\ \-i\ \-c\ sid\fP\fP\[CR]
.br
\f[CR]\ \ \[em]\[em]\[em]\ Chroot\ \[em]\[em]\[em]\fP
.br
\f[CR]\ \ Name\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ sid\fP
.br
\f[CR]\ \ Description\ \ \ \ \ \ \ \ \ \ \ \ Debian sid (unstable)\fP
.br
\f[CR]\ \ Type\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ plain\fP
.br
\f[CR]\ \ Priority\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ 3\fP
.br
\f[CR]\ \ Users\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ rleigh\fP
.br
\f[CR]\ \ Groups\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ sbuild\fP
.br
\f[CR]\ \ Root\ Users\fP
.br
\f[CR]\ \ Root\ Groups\ \ \ \ \ \ \ \ \ \ \ \ sbuild\fP
.br
\f[CR]\ \ Aliases\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ unstable\ unstable\-sbuild\ unstable\-p\fP
.br
\f[CR]owerpc\-sbuild\fP
.br
\f[CR]\ \ Environment\ Filter\ \ \ \ \ ^(BASH_ENV|CDPATH|ENV|HOSTALIASES|I\fP
.br
\f[CR]FS|KRB5_CONFIG|KRBCONFDIR|KRBTKFILE|KRB_CONF|LD_.*|LOCALDOMA\fP
.br
\f[CR]IN|NLSPATH|PATH_LOCALE|RES_OPTIONS|TERMINFO|TERMINFO_DIRS|TE\fP
.br
\f[CR]RMPATH)$\fP
.br
\f[CR]\ \ Run\ Setup\ Scripts\ \ \ \ \ \ true\fP
.br
\f[CR]\ \ Script\ Configuration\ \ \ script\-defaults\fP
.br
\f[CR]\ \ Session\ Managed\ \ \ \ \ \ \ \ true\fP
.br
\f[CR]\ \ Personality\ \ \ \ \ \ \ \ \ \ \ \ linux32\fP
.br
\f[CR]\ \ Location\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ /srv/chroot/sid\fP
.LP
Use \fI\-\-all\fP or \fI\-c\fP multiple times to use all or multiple chroots,
respectively.
.SS Running commands in a chroot
\f[CR]%\ \f[CB]schroot \-c sid /bin/ls\fP\fP\[CR]
.br
\f[CR][sid chroot] Running command: \[lq]/bin/ls\[rq]\fP
.br
\f[CR]CVS          sbuild\-chroot.c   sbuild\-session.h  schroot.conf.5\fP
.br
\f[CR]Makefile     sbuild\-chroot.h   schroot.1         schroot.conf.5.in\fP
.br
\f[CR]Makefile.am  sbuild\-config.c   schroot.1.in\fP
.br
\f[CR]Makefile.in  sbuild\-config.h   schroot.c\fP
.br
\f[CR]pam          sbuild\-session.c  schroot.conf\fP
.PP
\f[CR]%\ \f[CB]schroot \-c sid \-\- ls \-1 | head \-n 5\fP\fP\[CR]
.br
\f[CR][sid chroot] Running command: \[lq]ls \-1\[rq]\fP
.br
\f[CR]ABOUT\-NLS\fP
.br
\f[CR]AUTHORS\fP
.br
\f[CR]COPYING\fP
.br
\f[CR]ChangeLog\fP
.br
\f[CR]INSTALL\fP
.LP
Use \fI\-\-\fP to allow options beginning with \[oq]\-\[cq] or \[oq]\-\-\[cq]
in the command to run in the chroot.  This prevents them being interpreted as
options for schroot itself.  Note that the top line was echoed to standard
error, and the remaining lines to standard output.  This is intentional, so
that program output from commands run in the chroot may be piped and redirected
as required; the data will be the same as if the command was run directly on
the host system.
.SS Switching users
\f[CR]%\ \f[CB]schroot \-c sid \-u root\fP\fP\[CR]
.br
\f[CR]Password:\fP
.br
\f[CR][sid chroot] (rleigh\[->]root) Running login shell: \[lq]/bin/bash\[rq]\fP
.br
\f[CR]# \fP
.LP
If the user \[oq]rleigh\[cq] was in \f[CI]root\-users\fP in
\fI@SCHROOT_CONF@\fP, or one of the groups he belonged to was in
\f[CI]root\-groups\fP, they would be granted root access without

NEWS
NEWS                                                            -*- outline -*-
----

Welcome to schroot 1.5.1.  Please read these release notes carefully.

Full installation instructions are provided in the INSTALL file.  The
README file also contains more specific notes regarding building and
configuration.

* Major changes in 1.5.1:

  1) schroot no longer requires GCC to build, following the removal of
     GCC-specific features.  It may now build with other C++ compilers.

  2) Large file support is enabled by default.  This enables the use
     of files over 2 GiB in size on 32 bit architectures.

  3) The environment variable CHROOT_ALIAS has been added to the setup
     script execution environment, and SCHROOT_ALIAS_NAME to the user
     environment.  These may be used to conditionally alter behaviour
     depending upon the chroot alias used.

  4) dchroot and dchroot-dsa no longer use dchroot.conf.  Both
     programs now always use schroot.conf, and additionally use the
     same authentication mechanisms as schroot.  This is intended to
     provide the same basic configuration for all tools, and to also
     improve security by only having a single set of authentication
     rules.

* Major changes in 1.5.0:

  1) The deprecated options priority, run-setup-scripts and
     run-exec-scripts (all chroot types) and location ("plain" and
     "directory" chroot types) have been obsoleted and removed from
     the documentation.

  2) Source chroots no longer create a chroot with a -source suffix;
     the source: namespace should be used instead.  Likewise sessions
     are no longer present in the chroot: namespace, and are only
     found in the session: namespace.  This means the fully-qualified
     name must be used to refer to sessions except when performing
     actions which use session: as the default namespace.

  3) Support for zip file archives has been removed.  zip was not able
     to archive named pipes and device nodes, and so was not usable
     for chroot archival.

  4) AUTH_VERBOSITY is no longer set in setup scripts.  Please use
     VERBOSE instead.  VERBOSITY replaced and deprecated
     AUTH_VERBOSITY in version 1.4.5.

* Major changes in 1.4.23:

  None.

* Major changes in 1.4.22:

  1) Large file support is enabled by default.  This enables the use
     of files over 2 GiB in size on 32 bit architectures.

  2) Chroot profiles no longer bind filesystems with rbind.  Recursive
     bind mounting of /proc, /dev and /sys caused breakage with
     systemd due to its use of autofs mounts.  autofs interacts badly
     with bind mounting, leading to unmountable mount points.  While
     rbind is still possible, it is not done by default, and instead
     only specific filesystems are mounted; additional mounts required
     must be added to the profile fstab file.

  3) Session metadata includes the original chroot name.  This is
     available in the user environment as SCHROOT_CHROOT_NAME and
     in the setup scripts as CHROOT_NAME.

  4) A buildd profile is now provided.  This was previously provided
     by the sbuild package, but has now been consolidated with the
     main collection of schroot profiles.

  5) A helper program, schroot-sbuild has been provided in order to
     provide better privilege separation with sbuild.  This permits an
     authorised user in the sbuild group to run builds as the sbuild
     user.

* Major changes in 1.4.21:

  None.

* Major changes in 1.4.20:

  1) Add support for the Boost filesystem v3 library, to permit
     building with Boost version 1.46.  Older versions of Boost
     continue to be supported.

* Major changes in 1.4.19:

  None.

* Major changes in 1.4.18:

  None.

* Major changes in 1.4.17:

  None.

* Major changes in 1.4.16:

  1) Chroot naming restrictions introduced in 1.4.0 have been relaxed
     following a code audit to evaluate the security implications.
     The name may not contain a leading period (‘.’).  Any characters
     are permitted, with the following restrictions.  The characters
     ‘:’ (colon), ‘,’ (comma) and ‘/’ (forward slash) are not
     permitted anywhere in the name.  The name may also not contain a
     trailing tilde ('~').  See schroot.conf(5) for more information.

* Major changes in 1.4.15:

  None.

* Major changes in 1.4.14:

  None.

* Major changes in 1.4.13:

  None.

* Major changes in 1.4.12:

  None.

* Major changes in 1.4.11:

  None.

* Major changes in 1.4.10:

  None.

* Major changes in 1.4.9:

  None.

* Major changes in 1.4.8:

  1) Chroot names are now prefixed by a namespace.  Three namespaces
     are used in this release, "chroot:", "source:" and "session:" for
     chroots, source chroots and sessions, respectively.  These may
     all be selected with the --all-chroots, --all-source-chroots and
     --all-sessions options.  Individual chroots may be referred to
     with or without a prefix, depending upon the context.  For most
     actions, "chroot:" is the default prefix, while the
     --run-session, --recover-session and --end-session actions use
     "session:" as the default prefix.

  2) Source chroots previously used a -source suffix.  A chroot named
     "squeeze" providing a source chroot would also create a source
     chroot named "squeeze-source".  The source chroot is now named
     "source:squeeze", but a chroot with a -source suffix is still
     created for compatibility (actually now named
     "chroot:squeeze-source").  The -source suffix names will be
     dropped in the 1.5.x development releases and 1.6.x stable
     releases.

  3) Session chroots previously were in the same flat namespace as
     chroots.  Now that sessions are in a separate namespace, it is
     possible to create a session with the same name as the original
     chroot.  For example a chroot named "build" is actually
     "chroot:build" and so the session will be named "session:build".
     For compatibility session names are also still placed in the

schroot/schroot.c
/* schroot - securely enter a chroot
 *
 * Copyright © 2005  Roger Leigh <rleigh@debian.org>
 *
 * schroot is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * schroot is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston,
 * MA  02111-1307  USA
 *
 *********************************************************************/

#include <config.h>

#define _GNU_SOURCE
#include <locale.h>
#include <stdlib.h>
#include <stdio.h>

#include <syslog.h>

#include <glib.h>
#include <glib/gi18n.h>

#include "sbuild-chroot.h"
#include "sbuild-config.h"
#include "sbuild-session.h"

#include "schroot-options.h"

/**
 * print_version:
 * @file: the file to print to
 *
 * Print version information.
 */
void
print_version (FILE *file)
{
  g_fprintf(file, _("schroot (Debian sbuild) %s\n"), VERSION);
  g_fprintf(file, _("Written by Roger Leigh\n\n"));
  g_fprintf(file, _("Copyright (C) 2004-2005 Roger Leigh\n"));
  g_fprintf(file, _("This is free software; see the source for copying conditions.  There is NO\n"
		    "warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n"));
}

/**
 * get_chroot_options:
 * @config: an #SbuildConfig
 * @options: an #SchrootOptions
 *
 * Get a list of chroots based on the specified options (--all, --chroot).
 *
 * Returns a NULL-terminated string vector (GStrv).
 */
char **
get_chroot_options(SbuildConfig   *config,
		   SchrootOptions *options)
{
  char **chroots = NULL;

  if (options->all_chroots == TRUE || options->all_sessions == TRUE)
    {
      const GList *list = NULL;
      guint list_len = 0;

      list = sbuild_config_get_chroots(config);
      list_len = g_list_length((GList *) list);
      chroots = g_new(char *, list_len + 1);

      if (list != NULL)
	{
	  guint pos=0;
	  for (guint i=0; i < list_len; ++i)
	    {
	      GList *node = g_list_nth((GList *) list, i);
	      g_assert(node != NULL);
	      SbuildChroot *chroot = node->data;
	      g_assert(sbuild_chroot_get_name(chroot) != NULL);
	      gboolean active = sbuild_chroot_get_active(chroot);
 	      if ((active == FALSE && options->all_chroots == FALSE) ||
 		  (active == TRUE && options->all_sessions == FALSE))
		continue;
	      chroots[pos++] = g_strdup(sbuild_chroot_get_name(chroot));
	    }
	  chroots[pos] = NULL;
	}
    }
  else
    {
      char **invalid_chroots =
	sbuild_config_validate_chroots(config, options->chroots);

      if (invalid_chroots)
	{
	  for (guint i = 0; invalid_chroots[i] != NULL; ++i)
	    g_printerr(_("%s: No such chroot\n"), invalid_chroots[i]);
	  g_strfreev(invalid_chroots);
	  exit(EXIT_FAILURE);
	}
      g_strfreev(invalid_chroots);
      chroots = g_strdupv(options->chroots);
    }

  return chroots;
}

/**
 * debug_logfunc:
 * @log_domain: the log domain
 * @log_level: the logging level
 * @message: the message to log
 * @user_data: extra detail
 *
 * Log a debugging message.  This is a "NULL" message handler that
 * does nothing, discarding all messages.
 */
void debug_logfunc (const gchar *log_domain,
		    GLogLevelFlags log_level,
		    const gchar *message,
		    gpointer user_data)
{
  /* Discard all messages. */
}

/**
 * main:
 * @argc: the number of arguments
 * @argv: argument vector
 *
 * Main routine.
 *
 * Returns 0 on success, 1 on failure.
 */
int
main (int   argc,
      char *argv[])
{
  g_type_init();

  setlocale (LC_ALL, "");

  bindtextdomain (GETTEXT_PACKAGE, LOCALEDIR);
  textdomain (GETTEXT_PACKAGE);

#ifndef SBUILD_DEBUG
  /* Discard g_debug output for this logging domain. */
  g_log_set_handler (G_LOG_DOMAIN, G_LOG_LEVEL_DEBUG, debug_logfunc, NULL);
#endif

  openlog("schroot", LOG_PID|LOG_NDELAY, LOG_AUTHPRIV);

  /* Parse command-line options into opt structure. */
  SchrootOptions *options = schroot_options_parse(argc, argv);

  if (options->version == TRUE)
    {
      print_version(stdout);
      exit(EXIT_SUCCESS);
    }

  /* Initialise chroot configuration. */
  SbuildConfig *config = sbuild_config_new();
  g_assert (config != NULL);

  /* The normal chroot list is used when starting a session or running
     any chroot type or session, or displaying chroot information. */
  if (options->load_chroots == TRUE)
    sbuild_config_add_config_file(config, SCHROOT_CONF);
  /* The session chroot list is used when running or ending an
     existing session, or displaying chroot information. */
  if (options->load_sessions == TRUE)
    sbuild_config_add_config_directory(config, SCHROOT_SESSION_DIR);

  if (sbuild_config_get_chroots(config) == NULL)
    {
      if (options->quiet == FALSE)
	g_printerr(_("No chroots are defined in %s\n"), SCHROOT_CONF);
      exit (EXIT_FAILURE);
    }

  /* Print chroot list (including aliases). */
  if (options->list == TRUE)
    {
      sbuild_config_print_chroot_list(config, stdout);
      exit(EXIT_SUCCESS);
    }

  /* Get list of chroots to use */
  char **chroots = get_chroot_options(config, options);
  if (chroots == NULL)
    {
      g_printerr(_("The specified chroots are not defined in %s\n"), SCHROOT_CONF);
      exit (EXIT_FAILURE);
    }

  /* Print chroot information for specified chroots. */
  if (options->info == TRUE)
    {
      sbuild_config_print_chroot_info(config, chroots, stdout);
      exit (EXIT_SUCCESS);
    }

  if (options->session_operation == SBUILD_SESSION_OPERATION_BEGIN &&
      chroots[0] != NULL && chroots[1] != NULL)
    {
      g_printerr(_("Only one chroot may be specified when beginning a session\n"));
      exit (EXIT_FAILURE);
    }

  /* Create a session. */
  SbuildSession *session =
    sbuild_session_new("schroot",
		       config,
		       options->session_operation,
		       chroots);
  if (options->user)
    sbuild_auth_set_user(SBUILD_AUTH(session), options->user);
  if (options->command)
    sbuild_auth_set_command(SBUILD_AUTH(session), options->command);
  if (options->preserve)
    sbuild_auth_set_environment(SBUILD_AUTH(session), environ);
  sbuild_session_set_force(session, options->session_force);
  SbuildAuthVerbosity verbosity = SBUILD_AUTH_VERBOSITY_NORMAL;
  if (options->quiet)
    verbosity = SBUILD_AUTH_VERBOSITY_QUIET;
  else if (options->verbose)
    verbosity = SBUILD_AUTH_VERBOSITY_VERBOSE;
  sbuild_auth_set_verbosity(SBUILD_AUTH(session), verbosity);

  /* Set up authentication timeouts. */
  SbuildAuthConv *conv = SBUILD_AUTH_CONV(sbuild_auth_conv_tty_new());
  time_t curtime = 0;
  time(&curtime);
  sbuild_auth_conv_set_warning_timeout(conv, curtime + 15);
  sbuild_auth_conv_set_fatal_timeout(conv, curtime + 20);
  sbuild_auth_set_conv(SBUILD_AUTH(session), conv);

  /* Run session. */
  GError *session_error = NULL;
  sbuild_auth_run(SBUILD_AUTH(session), &session_error);
  if (session_error)
    g_printerr(_("Session failure: %s\n"), session_error->message);

  int exit_status = sbuild_session_get_child_status(session);

  g_object_unref(G_OBJECT(session));
  g_object_unref(G_OBJECT(config));
  g_strfreev(chroots);

  schroot_options_free(options);

  closelog();

  exit (exit_status);
}
